### PR TITLE
/usr/local/sbin/netwarn pop-ups (if iiab-network needs to be run) on MATE & RasPiOS with desktop + roles/network refactoring

### PR DESCRIPTION
**Pull request by holta**
1. Building on @jvonau's PR #3313 with network/tasks/install.yml syntax error fixed, cleaner pop-ups, ~and an in-line explanation for `#!/bin/sh` on the top line of scripts/netwarn.sh (despite the fact this 1st line is ignored on certain OS's!)~

2. Boot pop-ups tested on 'Raspberry Pi OS with desktop' on RPi 4.

3. I learned that `./iiab-network` is sometimes no longer enough &mdash; in such QA-like scenarios, consider running:

   ```
   ./runrole --reinstall network
   ```

4. Adding the same pop-ups to Mint and stock Ubuntu should hopefully now be extremely easy &mdash; in [roles/network/tasks/netwarn.yml](https://github.com/holta/iiab/blob/net_warn/roles/network/tasks/netwarn.yml)

5. _What testing or refinements are needed before merging this PR?_

Related:

- PR #2609
- PR #3308
- PR #3313

### CODE REVIEW

**Comment by holta** [comment]
> 4. Adding the same pop-ups to Mint and stock Ubuntu should hopefully now be extremely easy — in [roles/network/tasks/netwarn.yml](https://github.com/holta/iiab/blob/net_warn/roles/network/tasks/netwarn.yml)

Auto-launch of browser displaying http://box.lan [or http://box/home/] should also be turned on (in all above 3-4 graphical desktop OS's) &mdash; within [roles/www-options/tasks/main.yml](https://github.com/holta/iiab/blob/44af07638b9395207c6b72d604c93e4553be4f1a/roles/www_options/tasks/main.yml#L26-L48)

File: scripts/netwarn.sh [hunk 1]
```diff
@@ -1,29 +1,32 @@
...
-        ;;
-
-    0)
-        x-terminal-emulator -e /usr/local/bin/iiab-network
-        ;;
-esac
+zenity --question --width=360 --text="IIAB needs to configure networking:\n\n► Internet must be live before you begin.\n►You might be prompted for your password.\n\nContinue?  (This can take 2-3 minutes)"
+rc=$?
+if [ "$rc" != "0" ]; then
+    exit $rc
```
Comment (jvonau) [inline scripts/netwarn.sh]: The ! here is what messes with Ubuntu-Mate-lightDM... Unless you have tested this back off and create your own version.

**Comment by holta** [comment]
Tested on Ubuntu MATE on RPi 4.  Two glitches arose, that are probably related:

1. As a test if /etc/iiab/iiab.env does not contain `STAGE=9`, the boot pop-up displays a very unhelpful error 255 immediately, unfortunately hiding this critical message, that I've reproduced by doing it manually at the command-line:

   ```
   iiab-admin@box:~$ sudo iiab-network
   Reading /etc/iiab/iiab.env
   
   Extracted STAGE=1 (counter) from /etc/iiab/iiab.env
   
   EXITING: STAGE (counter) value == 1
   
   IIAB Stage 3 not complete.
   
   Please run: ./iiab-install
   iiab-admin@box:~$
   ```

2. Worse, even after setting `STAGE=9` in iiab.env and rebooting, which allows `iiab-network` to complete (and correctly sets /etc/iiab/install-flags/iiab-network-complete) this same pop-up arises at the END of the job too:

   iiab-network_popup_error_255_on_mate

Either way, the instructions to examine /opt/iiab/iiab/iiab-network.log are obviously incorrect (in fact this file /opt/iiab/iiab/iiab-network.log contains no errors, showing `failed=0` after iiab-network completes in its x-terminal-emulator window, every time.)

iiab-diagnostics: http://sprunge.us/iXWDPa?en

File: scripts/netwarn.sh [hunk 2]
```diff
@@ -1,29 +1,32 @@
...
+fi
 
-if [ "$?" = "1" ]; then
-    zenity --warning --width=350 --text="iiab-network exited with error.\n\nPlease review /opt/iiab/iiab/iiab-network.log"
-    exit 1
+x-terminal-emulator -e /usr/local/bin/iiab-network
+rc=$?
+if [ "$rc" != "0" ]; then
```
Comment (jvonau) [inline scripts/netwarn.sh]: and here again !

**Comment by jvonau** [comment]
What you are chasing the the use of ! in the comparison... Now you know... Stop breaking my code because you thing you know more..

File: scripts/netwarn.sh (hunk 1 above)
Comment (holta) [reply scripts/netwarn.sh]: `!=` is POSIX standard sh (dash) and definitely runs no problem on MATE, RasPiOS both.

Clicking 'No' or 'Yes' both work correctly (to progress to the next stage anyway!)

**Comment by jvonau** [comment]
https://github.com/holta/iiab/commit/85a5f8fd985d000ec1ddea47b0538e29d40cddd8 opps that was a copy and paste error, sorry about that one.. 

> Extracted STAGE=1 (counter) from /etc/iiab/iiab.env
EXITING: STAGE (counter) value == 1
IIAB Stage 3 not complete.
Please run: ./iiab-install

I'll assume you did not run iiab-install to actually complete the installation before moving on...

**Comment by jvonau** [comment]
Stop playing games you took a shortcut stage one with different [version] and [runtime] entries points to meddling, start with the clean version of Mate desktop without iiab installed or ask what the best way to test this is.. I'm out for the evening back in 12 hours. 
> 59 STAGE=1 
 65 =IIAB==========================================================================
  66 -rw-r--r-- 1 root root 2816 Jul 21 17:18 /etc/iiab/iiab.ini
  67 
  68 
  69 [location]
  70 iiab_base = /opt/iiab
  71 iiab_dir = /opt/iiab/iiab
  72 [version]
  73 distribution = Ubuntu
  74 arch = aarch64
  75 iiab_base_ver = 8.0
  76 iiab_remote_url = https://github.com/iiab/iiab
  77 iiab_branch = master
  78 iiab_commit = be360a67cf9f0e4ecf88738eca2e67b8b1c0facb
  79 iiab_recent_tag = 8.0-preview-2
  80 install_date = 2022-07-11T05:10:12Z
  81 rpi_model = none
  82 devicetree_model = none
  83 [runtime]
  84 iiab_stage = 9
  85 iiab_base_ver = 8.0
  86 iiab_revision = 0
  87 iiab_remote_url = none
  88 runtime_branch = net_warn
  89 runtime_commit = 44af07638b9395207c6b72d604c93e4553be4f1a
  90 iiab_recent_tag = 8.0-preview-2
  91 runtime_date = 2022-07-21T21:18:07Z
  92 ansible_version = 2.13.1
  93 kernel = 5.15.0-1012-raspi
  94 memory_mb = 1844
  95 swap_mb = 1023
  96 rpi_model = Raspberry Pi 4 Model B Rev 1.4
  97 devicetree_model = Raspberry Pi 4 Model B Rev 1.4
  98 first_run = False
  99 local_tz = EDT
 100 etc_localtime.stdout = America/New_York
 101 FQDN_changed = False

**Comment by holta** [comment]
x-terminal-emulator and mate-terminal [always return 255 on MATE boot](https://github.com/iiab/iiab/pull/3318#issuecomment-1191955300).

This PR's /usr/local/sbin/netwarn correctly [trapped the misleading] 255 "errors": https://github.com/holta/iiab/blob/44af07638b9395207c6b72d604c93e4553be4f1a/roles/network/templates/netwarn/netwarn#L22-L27

Until there's an alternative to mate-terminal with autostart (currently providing no meaningful error-checking, as 255 appears to be the result no matter what command is passed to to mate-terminal) I've added an interim patch to bypass x-terminal-emulator ~= mate-terminal[*] error-checking entirely on MATE.

[*] /usr/bin/x-terminal-emulator -> /etc/alternatives/x-terminal-emulator -> /usr/bin/mate-terminal.wrapper [PERL script] -> /usr/bin/mate-terminal

**Comment by holta** [comment]
1. Pop-ups now work on MATE and RasPiOS with desktop.

   - Why? Logic was introduced to intercept iiab-network's return code (error code) before mate-terminal's buggy return code (255 every time, when used with autostart) obliterates it.  The hack (temporarily recording to /tmp/iiab-network.rc) circumvents MATE's flaw, allowing legit iiab-network error-checking (return codes) on MATE.

2. iiab-network was strengthened to work with the above, recording essentially ALL of its errors to /opt/iiab/iiab/iiab-network.log (using tee, so these also appear when iiab-network is run at the command-line).

   - Conclusion: the pop-up asking you to check /opt/iiab/iiab/iiab-network.log is now Genuinely Meaningful.

3. The entire sh / dash / bash story was complete disinformation.

   - Why?  Simply put `#!/bin/bash` on top and everything works.  Enough said (far too much said).

**Comment by holta** [comment]
netplan.yml cannot run prior to detected_network.yml &mdash; as it needs the output of `ls /etc/netplan` in the `netplan` Ansible variable:

```
TASK [1-prep : Netplan in use on Ubuntu 18.04+] ************************************************************************************************************************
included: /opt/iiab/iiab/roles/network/tasks/netplan.yml for 127.0.0.1

TASK [1-prep : Disable cloud-init the easy way] ************************************************************************************************************************
fatal: [127.0.0.1]: FAILED! => {"msg": "'netplan' is undefined"}
```

@jvonau can you take a look?

http://sprunge.us/BwyLeO?en (iiab-diagnostics from Ubuntu Server 22.04 x86_64 VM)

**Comment by holta** [comment]
LARGE-sized smoke test running on 64-bit RasPiOS Lite on RPi 4 &mdash; FYI network/tasks/install.yml ran successfully and it's now running Stage 2:

```
root@raspberrypi:~# iiab-summary
install_date = 2022-07-22T20:03:53Z   Current TZ: UTC

iiab: 9f5032d7, 37 PR's / 207 commits since tag 8.0-preview-2
    "Comment out 'netplan apply' from restart.yml (trim netplan.yml later?)"
    https://github.com/holta/iiab   branch: net_warn

/etc/iiab/pr-list-pulled:
2022-07-22 20:02:03 UTC #3318 18 9f5032d7 https://github.com/holta/iiab net_warn "/usr/local/sbin/netwarn pop-ups (if iiab-network needs to be run) on MATE & RasPiOS with desktop + roles/network refactoring"

iiab-admin-console: beeae89, 44 PR's / 125 commits since tag v0.5.5.9
    "Merge pull request #506 from tim-moody/0.6.0.4-enh2"
    https://github.com/iiab/iiab-admin-console   branch: master

Raspberry Pi reference 2022-04-04
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 27a8050c3c06e567c794620394a8c2d74262a516, stage2
/etc/debian-version: 11.4
display-manager? inactive   Arch1: arm64   Arch2: armhf
5.15.32-v8+ #1538 SMP PREEMPT Thu Mar 31 19:40:39 BST 2022 unknown
Cortex-A72   CPU(s): 4   Mem: 1849 Swap: 1023
Raspberry Pi 4 Model B Rev 1.1
Temperature(s): 73036

Filesystem      Size  Used Avail Use% Mounted on
/dev/root        59G  2.9G   55G   5% /      ZIMs: 0 OER2Go: 0 Apps2B: 27

lo eth0 wlan0 tun0
openvpn_enabled: True
openvpn_handle: 182-rpi4-lite64-LARGE-PR3318
192.168.0.182 10.8.0.14
```

**Comment by holta** [comment]
WiFi hotspot works with above RasPiOS Lite on RPi 4 test:

http://sprunge.us/w0j4ji?en (iiab-diagnostics)

**Comment by holta** [comment]
Good Enough for now, after repeated testing yesterday and today:

An `/etc/profile.d/` version like /etc/local/sbin/netwarn for ssh sessions (across all OS's/distros/window managers) might also make sense in future?

**Comment by holta** [comment]
1) The 48 timestamps (below from-beginning-to-end-of-the-87 minutes, from 20:00 to 21:27 across this LARGE-sized install) help illustrate IIAB's install path.

2) Some of these ~49 executables/scripts might be pruned in the coming year, if they're no longer in use, or very rarely used.

```
root@box:~# ls -ltr /usr/local/*bin/iiab*
-rwxr-xr-x 1 root root 53 Jul 22 20:06 /usr/local/bin/iiab-network

# /usr/local/sbin/netwarn would also appear here, if OS had been 'RasPiOS with desktop' or MATE

root@box:~# ls -ltr /usr/*bin/iiab*
-rwxr--r-- 1 root root 31159 Jul 22 20:00 /usr/sbin/iiab
-rwxr-xr-x 1 root root  3648 Jul 22 20:04 /usr/bin/iiab-summary
-rwxr-xr-x 1 root root 13310 Jul 22 20:04 /usr/bin/iiab-diagnostics
-rwxr-xr-x 1 root root  9346 Jul 22 20:05 /usr/bin/iiab-support
-rwxr-xr-x 1 root root   996 Jul 22 20:05 /usr/bin/iiab-remote-on
-rwxr-xr-x 1 root root   949 Jul 22 20:05 /usr/bin/iiab-remote-off
lrwxrwxrwx 1 root root    21 Jul 22 20:05 /usr/bin/iiab-support-on -> /usr/bin/iiab-support
lrwxrwxrwx 1 root root    24 Jul 22 20:05 /usr/bin/iiab-support-off -> /usr/bin/iiab-remote-off
lrwxrwxrwx 1 root root    23 Jul 22 20:05 /usr/bin/iiab-vpn-on -> /usr/bin/iiab-remote-on
lrwxrwxrwx 1 root root    24 Jul 22 20:05 /usr/bin/iiab-vpn-off -> /usr/bin/iiab-remote-off
-rwxr-xr-x 1 root root  9869 Jul 22 20:06 /usr/bin/iiab-remoteit
-rwxr-xr-x 1 root root  1403 Jul 22 20:06 /usr/bin/iiab-remoteit-off
-rwxr-xr-x 1 root root  2295 Jul 22 20:06 /usr/bin/iiab-apps-to-be-installed
-rwxr-xr-x 1 root root   589 Jul 22 20:08 /usr/bin/iiab-internet-on
-rwxr-xr-x 1 root root   590 Jul 22 20:08 /usr/bin/iiab-internet-off
-rwxr-xr-x 1 root root  2853 Jul 22 20:08 /usr/sbin/iiab-expand-rootfs
-rwxr-xr-x 1 root root  4707 Jul 22 20:13 /usr/bin/iiab-refresh-wiki-docs
-rwxr-xr-x 1 root root   152 Jul 22 20:14 /usr/bin/iiab-bt-pan-on
-rwxr-xr-x 1 root root   245 Jul 22 20:14 /usr/bin/iiab-bt-pan-off
-rwxr-xr-x 1 root root    52 Jul 22 20:14 /usr/bin/iiab-bt-pan-discoverable-on
-rwxr-xr-x 1 root root    62 Jul 22 20:14 /usr/bin/iiab-bt-term-on
-rwxr-xr-x 1 root root    62 Jul 22 20:14 /usr/bin/iiab-bt-term-off
-rwxr-xr-x 1 root root   167 Jul 22 20:14 /usr/bin/iiab-usb_lib-show-all-on
-rwxr-xr-x 1 root root   168 Jul 22 20:14 /usr/bin/iiab-usb_lib-show-all-off
-rwxr-xr-x 1 root root   501 Jul 22 20:14 /usr/sbin/iiab-clean-usb.sh
-rwxr-xr-x 1 root root   876 Jul 22 20:50 /usr/bin/iiab-make-kiwix-lib
-rwxr-xr-x 1 root root  2807 Jul 22 20:50 /usr/bin/iiab-make-kiwix-lib.py
-rwxr-xr-x 1 root root  4449 Jul 22 21:01 /usr/bin/iiab-install-map-region
-rwxr-xr-x 1 root root 30324 Jul 22 21:01 /usr/bin/iiab-extend-sat.py
-rwxr-xr-x 1 root root   344 Jul 22 21:01 /usr/bin/iiab-summarize.sh
-rwxr-xr-x 1 root root  3133 Jul 22 21:01 /usr/bin/iiab-maps-finish.py
-rwxr-xr-x 1 root root  1738 Jul 22 21:01 /usr/bin/iiab-make-init.py
-rwxr-xr-x 1 root root   274 Jul 22 21:01 /usr/bin/iiab-update-map
-rwxr-xr-x 1 root root   397 Jul 22 21:02 /usr/bin/iiab-mongodb-repair-if-no-lock
-rwxr-xr-x 1 root root   284 Jul 22 21:12 /usr/sbin/iiab-divert-to-nginx
-rwxr-xr-x 1 root root   852 Jul 22 21:12 /usr/sbin/iiab-make-cp-servers.py
-rwxr-xr-x 1 root root  2591 Jul 22 21:20 /usr/bin/iiab-check-firmware
-rwxr-xr-x 1 root root   256 Jul 22 21:21 /usr/bin/iiab-hotspot-on
-rwxr-xr-x 1 root root   308 Jul 22 21:21 /usr/bin/iiab-hotspot-off
-rwxr-xr-x 1 root root  3095 Jul 22 21:21 /usr/sbin/iiab-test-wifi
-rwxr-xr-x 1 root root 11546 Jul 22 21:21 /usr/bin/iiab-gen-iptables
lrwxrwxrwx 1 root root    46 Jul 22 21:27 /usr/bin/iiab-update-menus -> /opt/admin/cmdsrv/scripts/iiab_update_menus.py
-rwxr-xr-x 1 root root   219 Jul 22 21:27 /usr/bin/iiab-get-kiwix-cat
-rwxr-xr-x 1 root root   234 Jul 22 21:27 /usr/bin/iiab-get-oer2go-cat
-rwxr-xr-x 1 root root   267 Jul 22 21:27 /usr/bin/iiab-install-oer2go-mod
-rwxr-xr-x 1 root root    90 Jul 22 21:27 /usr/bin/iiab-sync-menu-defs
-rwxr-xr-x 1 root root  1818 Jul 22 21:27 /usr/bin/iiab-cmdsrv-ctl
```

**Comment by jvonau** [comment]
Could of used a different exit code for each different exit condition so one would know with one caused the exiting.

**Pull request merged**